### ISSUE TITLE
-Wreturn-stack-address false positive

### ISSUE THREAD

**Issue by hokein** [llvm/llvm-project#127195]
https://godbolt.org/z/YW8eeWG1x

```
namespace std {
template <typename T>
struct basic_string_view {
  const T* data() const;
};
using string_view = basic_string_view<char>;
}  // namespace std

template <typename T>
struct StatusOr {
  T* operator->() [[clang::lifetimebound]];
};

const char* s() {
  StatusOr<std::string_view> s;

  return s->data(); // bogus -Wreturn-stack-adddress diagnostic.
}
```

**Comment by hokein** [comment]
This is a regression caused by https://github.com/llvm/llvm-project/pull/114044.
